lib/Rest/HttpRequest.cpp

```diff
@@ -415,11 +417,160 @@ void HttpRequest::parseHeader(size_t length) {
   }
 }
 
-void HttpRequest::setArrayValue(std::string const&& key, std::string const&& value) {
-  _arrayValues[key].emplace_back(value);
+namespace {
+  std::string url_decode(const char* begin, const char* end) {
+    std::string out;
+    out.reserve(static_cast<size_t>(end - begin));
+    for (const char* i = begin; i != end; ++i) {
+      std::string::value_type c = (*i);
+      if (c == '%') {
+        if (i + 1 != end && i + 2 != end) {
+          int h = StringUtils::hex2int(i[1], 0) << 4;
+          h += StringUtils::hex2int(i[2], 0);
+          out.push_back(static_cast<char>(h & 0x7F));
+          i += 2;
+        }
+      } else if (c == '+') {
+        out.push_back(' ');
+      } else {
+        out.push_back(c);
+      }
+    }
+    return out;
+  }
 }
 
-void HttpRequest::setArrayValue(char* key, size_t length, char const* value) {
+void HttpRequest::parseUrl(const char* url, size_t urlLen) {
+  _fullUrl.reserve(urlLen);
+  // get rid of '//'
+  for (size_t i = 0; i < urlLen; ++i) {
+    _fullUrl.push_back(url[i]);
+    if (url[i] == '/') {
+      while (i + 1 < urlLen && url[i+1] == '/') {
+        ++i;
+      }
+    }
+  }
+  
+  const char* start = _fullUrl.data();
+  const char* end = start + _fullUrl.size();
+  // look for database name in URL
+  if (_fullUrl.size() >= 5) {
+    char const* q = start;
+    
+    // check if the prefix is "_db"
+    if (q[0] == '/' && q[1] == '_' && q[2] == 'd' && q[3] == 'b' && q[4] == '/') {
+      // request contains database name
+      q += 5;
+      start = q;
+      
+      // read until end of database name
+      while (q < end) {
+        if (*q == '/' || *q == '?' || *q == ' ' || *q == '\n' || *q == '\r') {
+          break;
+        }
+        ++q;
+      }
+      
+      TRI_ASSERT(q >= start);
+      _databaseName = std::string(start, q - start);
+      
+      start = q;
+    }
+  }
+  
+  char const* q = start;
+  while (q != end && *q != '?') {
+    ++q;
+  }
+  
+  if (*q == '?' || q == end) {
```

Comment:
Reverse the conditions here, otherwise we may dereference `end` (which technically is ok here, but not good practice):
```suggestion
  if (q == end || *q == '?') {
```